### PR TITLE
Readymade Templates: Use live previews

### PR DESCRIPTION
**Pull request by dsas**
Related to https://github.com/Automattic/dotcom-forge/issues/8138

## Proposed Changes

Readymade Templates now use a live preview on the collection screen rather than a hard-coded screenshot.

## Why are these changes being made?

This ensures that the preview is always accurate and doesn't have to be manually curated.

## Testing Instructions

 1. Apply D155716-code to your sandbox
 2. Sandbox public-api
 3. Use the calypso live link to go to /patterns
 4. Scroll down and look at the readymade template section
 5. Note that they display the contents from the stored RT rather than being hardcoded to neonfit
 6. Click on a RMT and notice that preview works in the details page

props to @mmtr for his original live preview work, which I've butchered here.

Before | After
-------|------
 Screenshot 2024-07-19 at 14 26 19| Screenshot 2024-07-19 at 14 26 57

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by dsas** [comment]
Some of the previews still look off - bakery has the wrong background colour for example, investigating...

**Comment by BogdanUngureanu** [comment]
Hmm... besides the style issue, it seems that the details page is still displaying the hardcoded image: https://container-intelligent-gould.calypso.live/patterns/site-layouts/15742

**Comment by dsas** [comment]
> Hmm... besides the style issue, it seems that the details page is still displaying the hardcoded image: https://container-intelligent-gould.calypso.live/patterns/site-layouts/15742

I know, that's why the PR is "...in carousel", I've updated the description to be a bit clearer

**Comment by dsas** [comment]
> The section variation css should be added here: https://github.com/WordPress/gutenberg/blob/154f1aaeea73ba1ed2c17373814c8c3620dc9626/lib/block-supports/block-style-variations.php#L93C10-L93C63. That function is hooked on `render_block_data`. It outputs inline css if some of the blocks have an `is-style-SLUG` class and so doesn't use the style engine that `WPCOM_REST_API_V2_Endpoint_Block_Renderer_Patterns` currently extracts styles from.

I can't see where the chain following this is hookable in anyway 🤔 I think we either need to basically reimplement that method and return the styles differently (hoping it doesn't change often).........or do previews differently

**Comment by dsas** [comment]
There's another issue with this PR where it doesn't work with a clean local storage, yet it works if you then refresh the page. I assume it's related to an early return before the ready made templates are fetched 🤔 

> use-render-readymade-template.ts:21 Warning: React has detected a change in the order of Hooks called by ReadymadeTemplatesSection

```
   Previous render            Next render
   ------------------------------------------------------
1. useContext                 useContext
2. useState                   useState
3. useEffect                  useEffect
4. useMemo                    useMemo
5. useRef                     useRef
6. useState                   useState
7. useState                   useState
8. useEffect                  useEffect
9. undefined                  useContext
```

**Comment by rcrdortiz** [comment]
I'm seeing empty previews, not sure if I'm doing something wrong 🤔. I've applied the patch and have my API sandboxed
Screenshot 2024-07-22 at 15 30 56

**Comment by BogdanUngureanu** [comment]
@rcrdortiz Try cleaning the App state (DevTools > Application > Storage). React-query has a 5-minute cache on API requests.

File: client/my-sites/patterns/hooks/use-render-readymade-template.ts [hunk 1]
```diff
@@ -1,23 +1,41 @@
...
+			const x = wpcom.req.post(
 				{
-					path: `/sites/${ RENDERER_SITE_ID }/block-renderer/patterns/render`,
+					path: `/sites/${ RENDERER_SITE_ID }/block-renderer/content/render`,
 					apiNamespace: 'wpcom/v2',
+					method: 'POST',
 				},
 				{
-					pattern_ids: readymadeTemplate.patterns.map( ( { id } ) => id ).join( ',' ),
 					site_title: readymadeTemplate.title,
```
Comment (mmtr) [inline client/my-sites/patterns/hooks/use-render-readymade-template.ts:17]: This site title plan seems to be ignored. All the previews are displaying "Assembler" as site title.

**Comment by mmtr** [comment]
Overall, this works really well! Although there are a few differences when I compare the previews with how my sites actually look in the frontend:

Preview | Frontend | Differences
--- | --- | ---
Screenshot 2024-07-25 at 11 33 51 | Screenshot 2024-07-25 at 11 29 13 | "Just breathe" is bold in the preview, and non-bold in the frontend
Screenshot 2024-07-25 at 11 33 54 | Screenshot 2024-07-25 at 11 30 31 | "View projects" is yellow on the preview, white/transparent on the frontend
Screenshot 2024-07-25 at 11 33 57 | Screenshot 2024-07-25 at 11 32 51 | There a gap between the header and the content in the frontend which is not in the preview
Screenshot 2024-07-25 at 11 34 13 | Screenshot 2024-07-25 at 11 39 42 | There a gap between the header and the content in the frontend which is not in the preview
Screenshot 2024-07-25 at 11 34 19 | Screenshot 2024-07-25 at 11 41 13 | The site title is way bigger in the preview than in the frontend

File: client/my-sites/patterns/components/readymade-template-details/client.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+	return (
+		<ReadymadeTemplateRenderer
+			renderedReadymadeTemplate={ renderedReadymadeTemplate }
+			viewportWidth={ 1200 }
+		/>
+	);
+};
+
+export const ReadymadeTemplateDetailsClient = ( { id }: { id: number } ) => {
```
Comment (mmtr) [inline client/my-sites/patterns/components/readymade-template-details/client.tsx]: I think you can use the `ReadymadeTemplateDetailsFC` type here.

File: client/my-sites/patterns/components/readymade-template-details/server.tsx [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { ReadymadeTemplateDetails } from 'calypso/my-sites/patterns/components/readymade-template-details/index';
+
+export const ReadymadeTemplateDetailsServer = ( { id }: { id: number } ) => {
```
Comment (mmtr) [inline client/my-sites/patterns/components/readymade-template-details/server.tsx]: I think you can use the `ReadymadeTemplateDetailsFC` type here.

Comment (mmtr) [reply client/my-sites/patterns/components/readymade-template-details/server.tsx]: Never mind, just seen `ReadymadeTemplateDetailsFC` is now used for the inner component.

File: client/my-sites/patterns/components/readymade-template-details/client.tsx (hunk 2 above)
Comment (mmtr) [reply client/my-sites/patterns/components/readymade-template-details/client.tsx]: Never mind, just seen `ReadymadeTemplateDetailsFC` is now used for the inner component.

File: client/my-sites/patterns/components/readymade-template-details/server.tsx [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { ReadymadeTemplateDetails } from 'calypso/my-sites/patterns/components/readymade-template-details/index';
```
Comment (mmtr) [inline client/my-sites/patterns/components/readymade-template-details/server.tsx]: `index` is unnecessary:
```suggestion
import { ReadymadeTemplateDetails } from 'calypso/my-sites/patterns/components/readymade-template-details';
```

File: client/my-sites/patterns/components/readymade-template-details/client.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import { useRenderReadymadeTemplate } from 'calypso/my-sites/patterns/hooks/use-render-readymade-template';
+import { ReadymadeTemplate } from 'calypso/my-sites/patterns/types';
+import { ReadymadeTemplateDetails } from './index';
+
+const ReadymadeTemplatePreview = ( {
+	readymadeTemplate,
+}: {
+	readymadeTemplate: ReadymadeTemplate;
+} ) => {
+	const { data: renderedReadymadeTemplate = {} } = useRenderReadymadeTemplate( readymadeTemplate );
```
Comment (mmtr) [inline client/my-sites/patterns/components/readymade-template-details/client.tsx:12]: Can we move `useRenderReadymadeTemplate` inside `ReadymadeTemplateRenderer`?

The `ReadymadeTemplatePreview` components looks like an unnecessary wrapper.

File: client/my-sites/patterns/index.node.tsx [hunk 6]
```diff
@@ -101,7 +103,7 @@ function renderReadymadeTemplateDetails( context: RouterContext, next: RouterNex
 
 	context.primary = (
 		<PatternsWrapper>
-			<ReadymadeTemplateDetails id={ parseInt( context.params.id ) } />
+			<ReadymadeTemplateDetailsServer id={ parseInt( context.params.id ) } />
```
Comment (mmtr) [inline client/my-sites/patterns/index.node.tsx]: Considering that `ReadymadeTemplateDetailsServer` and `ReadymadeTemplateDetails` are the exact the same component, I wonder if we can remove `ReadymadeTemplateDetailsServer` and just use `ReadymadeTemplateDetails` here.

File: client/my-sites/patterns/index.node.tsx [hunk 7]
```diff
@@ -41,6 +42,7 @@ function renderPatterns( context: RouterContext, next: RouterNext ) {
 				<PatternLibrary
 					categoryGallery={ CategoryGalleryServer }
 					patternGallery={ PatternGalleryServer }
+					readymadeTemplates={ ReadymadeTemplatesServer }
```
Comment (mmtr) [inline client/my-sites/patterns/index.node.tsx]: Considering that `ReadymadeTemplatesServer` and `ReadymadeTemplatesSection` are the exact the same component, I wonder if we can remove `ReadymadeTemplatesServer` and just use `ReadymadeTemplatesSection` here.

File: client/my-sites/patterns/hooks/use-render-readymade-template.ts [hunk 8]
```diff
@@ -1,23 +1,41 @@
...
+		.replace( /^-+|-+$/g, '' ); // remove runs of dashes and trailing dashes
+}
+
 export const useRenderReadymadeTemplate = ( readymadeTemplate: ReadymadeTemplate ) =>
-	useQuery< { [ key: string ]: RenderedPattern } >( {
+	useQuery( {
 		queryKey: [ 'pattern-library', 'readymade-template', readymadeTemplate.template_id, 'render' ],
 		queryFn() {
-			return wpcom.req.get(
+			const x = wpcom.req.post(
```
Comment (mmtr) [inline client/my-sites/patterns/hooks/use-render-readymade-template.ts]: This `x` const is never used, can we simply return here?

File: client/my-sites/patterns/hooks/use-render-readymade-template.ts [hunk 9]
```diff
@@ -1,23 +1,41 @@
-import { RenderedPattern } from '@automattic/block-renderer';
 import { useQuery } from '@tanstack/react-query';
 import wpcom from 'calypso/lib/wp';
 import { RENDERER_SITE_ID } from 'calypso/my-sites/patterns/constants';
 import type { ReadymadeTemplate } from 'calypso/my-sites/patterns/types';
 
+// Shouldnt have to do this.
```
Comment (mmtr) [inline client/my-sites/patterns/hooks/use-render-readymade-template.ts]: Right, we shouldn't need this. The API already returns _slugified_ style variations. Why do we need this?

File: client/my-sites/patterns/hooks/use-render-readymade-template.ts (hunk 1 above)
Comment (BogdanUngureanu) [reply client/my-sites/patterns/hooks/use-render-readymade-template.ts:17]: Nice catch, @mmtr! It was a problem in the API. I've updated the diff and now the site name is working properly.

File: client/my-sites/patterns/components/readymade-template-details/client.tsx (hunk 5 above)
Comment (BogdanUngureanu) [reply client/my-sites/patterns/components/readymade-template-details/client.tsx:12]: I'm getting an ESLint error:
```
ESLint: 'calypso/my-sites/patterns/types' import is restricted from being used by a pattern.(no-restricted-imports)

```

Comment (BogdanUngureanu) [reply client/my-sites/patterns/components/readymade-template-details/client.tsx:12]: Let's leave it as it is for now. :)

Comment (mmtr) [reply client/my-sites/patterns/components/readymade-template-details/client.tsx:12]: I guess `ReadymadeTemplateRenderer` shouldn't be part of the `@automattic/block-renderer` package.

File: client/my-sites/patterns/hooks/use-render-readymade-template.ts (hunk 9 above)
Comment (BogdanUngureanu) [reply client/my-sites/patterns/hooks/use-render-readymade-template.ts]: Not sure. Maybe Dean was getting a stale payload from the API. 🤷 

I've removed it anyway.

**Review by mmtr — APPROVED** [review]
Nice job!

**Pull request merged**